File: c#/Asset/Folders/CreateFolder.cs
/*
   CreateFolder.cs

   Marketo REST API Sample Code
   Copyright (C) 2016 Marketo, Inc.

   This software may be modified and distributed under the terms
   of the MIT license.  See the LICENSE file for details.
*/
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net;
using System.IO;
using System.Collections;
using Newtonsoft.Json;

namespace Samples
{

    class CreateFolder
    {
        private String host = "CHANGE ME"; //host of your marketo instance, https://AAA-BBB-CCC.mktorest.com
        private String clientId = "CHANGE ME"; //clientId from admin > Launchpoint
        private String clientSecret = "CHANGE ME"; //clientSecret from admin > Launchpoint
        public String name;//name of folder to create, required
        public Dictionary<String, dynamic> parent; //parent folder to create with two members, id and Type(Folder or Program), required
        public String description;//optional description of new folder

        /*
        public static void Main(string[] args)
        {
            var folder = new CreateFolder();
            folder.name = "New Folder CSharp";
            folder.parent = new Dictionary<string,dynamic>();
            folder.parent.Add("id", 5565);
            folder.parent.Add("type", "Folder");
            String result = folder.postData();
            Console.WriteLine(result);
        }
        */

        public String postData()
        {
            //Assemble the URL
            String url = host + "/rest/asset/v1/folders.json?access_token=" + getToken();
            //Form Encode the data
            String requestBody = bodyBuilder();
            Console.WriteLine(requestBody);
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = "POST";
            request.ContentType = "application/x-www-form-urlencoded";
            request.Accept = "application/json";
            StreamWriter wr = new StreamWriter(request.GetRequestStream());
            wr.Write(requestBody);
            wr.Flush();
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            return reader.ReadToEnd();
        }
        private String bodyBuilder()
        {
            var sb = new StringBuilder();
            sb.Append("&name=" + name);
            sb.Append("&parent=" + JsonConvert.SerializeObject(parent));
            if (description != null)
            {
                sb.Append("&description=" + description);
            }
            return sb.ToString();
        }

        private String getToken()
        {
            String url = host + "/identity/oauth/token?grant_type=client_credentials&client_id=" + clientId + "&client_secret=" + clientSecret;
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.ContentType = "application/json";
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            String json = reader.ReadToEnd();
            //Dictionary<String, Object> dict = JavaScriptSerializer.DeserializeObject(reader.ReadToEnd);
            Dictionary<String, String> dict = JsonConvert.DeserializeObject<Dictionary<String, String>>(json);
            return dict["access_token"];
        }
    }
}

File: php/Asset/Folders/CreateFolder.php
<?php
/*
   CreateFolder.php

   Marketo REST API Sample Code
   Copyright (C) 2016 Marketo, Inc.

   This software may be modified and distributed under the terms
   of the MIT license.  See the LICENSE file for details.
*/
$folder = new CreateFolder();
$parent = new stdClass();
$parent->id = 5566;
$parent->type = "Folder";
$folder->parent = $parent;
$folder->name = "New Folder - PHP";
print_r($folder->postData());

class CreateFolder{
	private $host = "CHANGE ME";
	private $clientId = "CHANGE ME";
	private $clientSecret = "CHANGE ME";
	public $name;//name of folder to create
	public $parent;//JSON object representing a folder with two members, id and type(Folder or Program)
	public $description;//optional description for new folder
	
	public function postData(){
		$url = $this->host . "/rest/asset/v1/folders.json?access_token=" . $this->getToken();
		$ch = curl_init($url);
		$requestBody = $this->bodyBuilder();
		curl_setopt($ch,  CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array('accept: application/json'));
		curl_setopt($ch, CURLOPT_POST, 1);
		curl_setopt($ch, CURLOPT_POSTFIELDS, $requestBody);
		curl_getinfo($ch);
		$response = curl_exec($ch);
		return $response;
	}
	
	private function getToken(){
		$ch = curl_init($this->host . "/identity/oauth/token?grant_type=client_credentials&client_id=" . $this->clientId . "&client_secret=" . $this->clientSecret);
		curl_setopt($ch,  CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array('accept: application/json',));
		$response = json_decode(curl_exec($ch));
		curl_close($ch);
		$token = $response->access_token;
		return $token;
	}
	private function bodyBuilder(){
		$requestBody = "";
		$requestBody .= "name=" . $this->name;
		$requestBody .= "&parent=" . json_encode($this->parent);
		if (isset($this->description)){
			$requestBody .= "&description=" . $this->description;
		}
		return $requestBody;
	}	
}